README
helpdeco -- utility program to dissect Windows help files
Copyright (C) 1996, 1997 Manfred Winterhoff
Copyright (C) 2001 Ben Collver
Copyright (C) 2005 Paul Wise

This file is part of helpdeco; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation,
Inc., 59 Temple Place, Suite 330, Boston, MA, 02111-1307, USA or visit:
http://www.gnu.org

helpdeco has been relicenced to the GNU GPL with permission from Manfred:
> ------------------------------------------------------------------------
> From: 	Manfred Winterhoff <mawin@gmx.net>
> Date: 	Sun, 3 Apr 2005 20:47:46 +0200
> 
> Oh, so you developed CHMDeco. So you will know what demand is on the
> Linux side to handle Windows Help files. I can not believe that the
> NetBSD port of HelpDeco does have any audience, but if you want to port
> HelpDeco to Linux, just do it, add your copyright to mine, and release
> your port with source under GPL, so that Debian and all other Linux 
> distributers can include it if they want.
> ------------------------------------------------------------------------

HELPDECO dissects HLP help files of Windows 3.0, 3.1, 3.11, and '95 and
many MVB multi media viewer titles into all files required for a rebuild
using the appropriate help compiler HC30, HC31, HCP, HCW, HCRTF, WMVC,
MMVC or MVC:
HPJ - help project file, use as parameter when calling help compiler
MVP - multi media project file, parameter for multi media help compiler
RTF - text file containing whole content of help file and all footnotes
PH  - phrases file (same as produced by help compiler)
ICO - icon of help file if embedded
BMP/WMF/SHG/MRB - embedded pictures in appropriate format
Baggage - all baggage files contained in help file

HELPDECO
Displays usage

HELPDECO helpfilename
Decompiles help file into all sources needed for a rebuild. All files are
created in current directory (should be empty). Existing files will be
overwritten without asking if option -y was specified.

Options may be appended to the command, separated using blanks, as follows:

-m
May be used to stop parsing macros for topic names. Help compiler will emit
Warning 4131.

-b
May be used to stop resolving browse sequences. Help source file than
contains no + footnotes.

-g
Turns off time consuming guessing of context names. Applicable when
HELPDECO reports 'no context ids found' or when you are not interested in
legible context ids. This doesn't affect functionality.
During guessing HELPDECO tries to reconstruct context ids from topic titles
and keywords. This may produce good results on some help files and no result
at all on others, depending on autoring tool used or working method of help
author.

-i
If you want to see which context ids are guessed by HELPDECO...

-hprefix
HELPDECO tries to guess context ids from topic titles and keywords, possibly
prefixed by idh_ or helpid_. If the help file was created using other
prefixes, they may be added to the table using the -h prefix option.

-a [annotationfilename.ANN]
Adds all annotations from annotationfile as annotations of user ANN into
RTF file. Default annotationfilename is helpfilename.ANN.

-s n
Splits output every n pages into a separate RTF file. For example
  HELPDECO helpdemo -s 100
decompiles helpdeco.hlp and creates different RTF files helpdem1.rtf,
helpdem2.rtf,.. each containing no more than 100 topics.

HELPDECO helpfilename -r
Converts help file into RTF file of same appearance if loaded into WinWord
as if displayed by WinHelp. To print or work with complete content. Specify
additional option -n, if no page breaks should separate topics.

HELPDECO helpfilename -c
Generates a *.CNT file used by WinHlp32, containing all chapters that have
titles assigned in the order they appear in the helpfile. This file should
then be edited using HCW 4.00 or any text editor into a hierarchical order.

HELPDECO helpfilename -e
Lists all references to external help files.

HELPDECO helpfilename -e -f
Lists all references to external help files and titles of topics that
contained these references.

HELPDECO helpfilename -p
Checks references to external help files. Referenced help file need to be
available to HELPDECO. This call doesn't produce any new files and doesn't
modify existing files. Errors are reported to stdout (screen).

HELPDECO helpfilename -d
Displays internal directory of help file. You may supply a *.MVB,*.M??,
*.ANN,*.CAC,*.AUX file instead of a *.HLP file.

HELPDECO helpfilename -x
Displays hex dump of internal directory

HELPDECO helpfilename "internalfilename"
Displays internal file in appropriate format if known, else hex dump. You
may redirect output into a file using >outfile.

HELPDECO helpfilename "internalfilename" -x
Displays hex dump of internal file

HELPDECO helpfilename "internalfilename" -x -t offset
Displays hex dump of internal file starting at offset, which may be
specified in decimal or preceeded with 0x in hex.

HELPDECO helpfilename "internalfilename" filename
Exports internal file into filename

*.ANN, *.CAC, *.AUX
These files are formatted like helpfiles, but HELPDECO can only be used to
display their internal directory or display or export internal files.

HELPDECO was written by
Manfred Winterhoff <mawin@gmx.net>, Geschw.-Scholl-Ring 17, 38444 Wolfsburg, Germany

If you have more questions, problems (or answers!), please feel free to
send me an EMail to mawin@gmx.net

HELPDECO is based upon HELPDUMP from Pete Davis published in:
The Windows Help File Format, Dr. Dobbs Journal, Sep/Oct 1993
Thanks to Holger Haase, who did a lot of work on picture file formats.
And thanks to Jürgen Müller for pointing out 32-bit differences.
Thanks to Bent Lynggaard for the information on help file free lists
and his contribution on context id guessing.

See HELPFILE.TXT for a description of the Windows help file format
as parsed by HELPDECO and related file formats like MRB/SHG.

The newest public version of HELPDECO is always available at:
http://sourceforge.net/projects/helpdeco/

The use of parts of decompiled help files may constitute a violation of
copyright law.


NEWS
helpdeco -- utility program to dissect Windows help files

Version 2.1.4:
* Fix Debian #393787: Use only one option character and default to using dash
  WARNING: This breaks command-line compatibility with helpdeco <= 2.1.3
* Fix various memory leaks and segfaults
* Use data structures in endian and alignment-safe way, fixing 64-bit issues
* Fix segfault in strupr & strlwr implementations, fixes bug ticket 4
* Minor portability fixes

Version 2.1.3:
* Add patches from Dirk Jagdmann to fix struct packing in GCC >= 3,
  use POSIX path lengths instead of DOS ones, and fix strlwr/strupr
* Improve const correctness in the source code

Version 2.1.2:
* Add GCC 4.0 patch from Andreas Jochens <aj@andaco.de>

Version 2.1.1:
* Relicence to the GNU General Public Licence >= 2
* Incorporate the NetBSD patches by Ben Collver
* Add a Makefile (by Paul Wise)

Version 2.1:
* macros may contain rtf meta characters
* Win95 topic names may contain (nearly) all characters
* 0x8000 in SWin->Maximize allowed
* ForeHelp creates PhrIndex/PhrImage _and_ _empty_ Phrases file
* HCRTF complained about > footnotes before # footnotes
* macro parsing changed again

Version 2.0:
* Guess context ids from titles and keywords based on idea of Bent Lynggaard
* Recompiled 16 bit EXE without register calling convention. BC++ 3.1 bug.
* Doesn't print last (stray) topic of HC30 help files
* Can list entry points into this help file (option /l)

Version 1.9: faster & better than ever...
* changed TopicPos, TopicOffset, Keyword maintenance
* changed unhash to 40 bit integer arithmetic
* fixed keyword footnotes [Bent Lynggaard]
* no [ALIAS] in MVP files
* no hidden text in option /r RTF files
* some changes in font and stylesheet handling
* corrects rounding error of HC31 on negative values
* handles non-underlined topic jumps

Version 1.8: used some spare days to clean up the to-do list...
* better tracking of TopicOffset during decompilation
* lists and checks references to external files, shows referencing topics
* can add annotations from .ANN file to decompiled .RTF file
* fixed bug in handling of pictures containing JumpId-macro hotspots
* changed parsing of macros (3rd attempt to guess what Microsoft did)
* fixed bug in popup/jump to external file / secondary window
* fixed bug in > footnote / |VIOLA internal file handling
* fixed bug in keyword assignment
* now removes LZ77 compression from exported SHGs/MRBs
* recreates Win 95 (HCW 4.00) [MACROS] section from internal |Rose file
* 32 bit version available
* handles LANGUAGE, [CHARTAB] and [GROUP] section of media view files

Version 1.7
* removed unnecessary output statement

Version 1.6 can now check references to external help files plus:
* duplicate macro names preceeding picture hotspot info skipped
* does not write Win95 commands to multi-media help project files
* changed unhash to circumvent Microsoft-C++ float rounding error
* handles keywords defined inside topic text

Version 1.5
* fixed static on buffer of TopicName function (affected HC30 files)

Version 1.4 fixes some bugs reported by different users:
* buffer overflow in expanding LZ77&RunLen (byPacked 3) images fixed
* embedded images {bmxwd} larger than 32k supported
* extract topic names from jump into external file if no file specified
* handles more phrases on HCRTF generated (Win95) help files
* Windows 3.1 (HC31) |Phrases always Zeck compressed
* LinkData2 buffer enlarged 1 byte to store trailing NUL character

Version 1.3
* parses examples of {bmc} etc. statements contained in help text correctly
* can now generate a *.CNT content file for Windows 95 / WinHlp32
* Microsoft C: ctype macros (isalnum/isprint) don't work with signed char

Version 1.2 fixes some severe bugs introduced in version 1.1 and:
* tells you which help compiler to use
* collects multiple keyword footnotes into single lines
* handles \r\n in COPYRIGHT
* converts SPC-macro (but only in [CONFIG] section)
* does not generate duplicate MAP-statements if possible
* {button} and {mci,mci_left,mci_right} commands supported
* [BITMAP]-section in HCRTF help files irritated transparent bitmaps

Version 1.1 now supports more features of Win95/HCRTF 4.00/WinHlp32:
* Supports LCID, CHARSET, AUTO-SIZE HEIGHT, CNT, INDEX_SEPARATORS
* Additional Win95 Macros (to extract original topic names)
* [CONFIG:n] of Win95 supported (internal file |CFn)
* Secondary windows with > footnote supported (internal file |VIOLA)
* Transparent bitmaps supported (bmct,bmlt,bmrt)
* Expanded internal limits as HCRTF allows larger items
* Now does RunLen compressed device dependend bitmaps
* Bugs in handling of metafiles removed
* Bug in placement of pack(1) removed
* Parsing of macros changed (is it really better now ?)


splitmrb.c
/*
helpdeco -- utility program to dissect Windows help files
Copyright (C) 1996 Manfred Winterhoff
Copyright (C) 2001 Ben Collver

This file is part of helpdeco; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation,
Inc., 59 Temple Place, Suite 330, Boston, MA, 02111-1307, USA or visit:
http://www.gnu.org
*/

/*
SPLITMRB - splits MRB into SHG/BMP/WMF and SHG into BMP/WMF files - Version 1.5
M.Winterhoff <mawin@gmx.net>, Geschw.-Scholl-Ring 17, 38444 Wolfsburg, Germany

usage: SPLITMRB filename[.mrb] ...
       SPLITMRB filename.shg ...

Output files are created in the current directory using resolution-dependent
extensions *.EGA,*.VGA,*.CGA,*.854,*.MAC or *.BMP,*.WMF or *.Snn,*.nnn (where
n is a digit from 0 to 9). Discarded hotspot info will be written to stdout.
*/
#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <limits.h>

typedef enum {FALSE,TRUE} BOOL;

typedef struct tagBITMAPFILEHEADER
{
    unsigned short bfType;
    unsigned long bfSize;
    unsigned short bfReserved1;
    unsigned short bfReserved2;
    unsigned long bfOffBits;
}
BITMAPFILEHEADER;

typedef struct tagBITMAPINFOHEADER
{
    unsigned long biSize;
    long biWidth;
    long biHeight;
    unsigned short biPlanes;
    unsigned short biBitCount;
    unsigned long biCompression;
    unsigned long biSizeImage;
    long biXPelsPerMeter;
    long biYPelsPerMeter;
    unsigned long biClrUsed;
    unsigned long biClrImportant;
}
BITMAPINFOHEADER;

typedef struct tagRECT
{
    short left;
    short top;
    short right;
    short bottom;
}
RECT;

typedef struct tagAPMFILEHEADER
{
    unsigned long dwKey;
    unsigned short hMF;
    RECT rcBBox;
    unsigned short wInch;
    unsigned long dwReserved;
    unsigned short wChecksum;
}
APMFILEHEADER;

unsigned short GetCWord(FILE *f)
{
    unsigned char b;

    b=getc(f);
    if(b&1) return (((unsigned short)getc(f)<<8)|(unsigned short)b)>>1;
    return ((unsigned short)b>>1);
}

unsigned long GetCDWord(FILE *f)
{
    unsigned short w;

    w = getw(f);
    if(w&1) return (((unsigned long)getw(f)<<16)|(unsigned long)w)>>1;
    return ((unsigned long)w>>1);
}

unsigned long GetDWord(FILE *f)
{
    unsigned short w;

    w=getw(f);
    return ((unsigned long)getw(f)<<16)|(unsigned long)w;
}

long copy(FILE *f,long bytes,FILE *out)
{
    long length;
    int size;
    static char buffer[1024];

    for(length=0;length<bytes;length+=size)
    {
	size=(int)(bytes-length>sizeof(buffer)?sizeof(buffer):bytes-length);
	fread(buffer,1,size,f);
	fwrite(buffer,size,1,out);
    }
    return length;
}

signed char count; /* for run len decompression */

int derun(int c,FILE *f) /* expand runlen compressed data */
{
    int i;

    if(count&0x7F)
    {
	if(count&0x80)
	{
	    putc(c,f);
	    count--;
	    return 1;
	}
	for(i=0;i<count;i++)
	{
	    putc(c,f);
	}
	count=0;
	return i;
    }
    count=(signed char)c;
    return 0;
}

long decompress(int method,FILE *f,long bytes,FILE *fTarget)
{
    static unsigned char lzbuffer[0x1000];
    int (*Emit)(int c,FILE *f);
    unsigned char bits,mask;
    int pos,len,back;
    long n;

    n=0;
    if(method&1)
    {
	Emit=derun;
	count=0;
    }
    else
    {
	Emit=fputc;
    }
    if(method&2)
    {
	mask=0;
	pos=0;
	while(bytes-->0L)
	{
	    if(!mask)
	    {
		bits=getc(f);
		mask=1;
	    }
	    else
	    {
		if(bits&mask)
		{
		    if(bytes--==0) break;
		    back=getw(f);
		    len=((back>>12)&15)+3;
		    back=pos-(back&0xFFF)-1;
		    while(len-->0)
		    {
			n+=Emit(lzbuffer[pos++&0xFFF]=lzbuffer[back++&0xFFF],fTarget);
		    }
		}
		else
		{
		    n+=Emit(lzbuffer[pos++&0xFFF]=getc(f),fTarget);
		}
		mask<<=1;
	    }
	}
    }
    else
    {
	while(bytes-->0L) n+=Emit(getc(f),fTarget);
    }
    return n;
}

int GetPackedByte(FILE *f) // RulLen decompression
{
    static unsigned char count,value;

    if(!f) // initialize
    {
	count=value=0;
    }
    else
    {
	if((count&0x7F)==0)
	{
	    count=getc(f);
	    value=getc(f);
	}
	else if(count&0x80)
	{
	    value=getc(f);
	}
	count--;
    }
    return value;
}

#if defined(__GNUC__) && __GNUC__ >= 3
#pragma pack (push, 1)
#endif
typedef struct
{
  uint8_t c1,c2,c3;
  uint16_t x,y,w,h;
  uint32_t hash;
} HOTSPOT
#if defined(__GNUC__) && __GNUC__ < 3
__attribute__((packed))
#endif
     ;
#if defined(__GNUC__) && __GNUC__ >= 3
#pragma pack (pop)
#endif

void PrintHotspotInfo(FILE *f)
{
    int i,l,n;
    HOTSPOT *hotspot;
    char name[80];
    char buffer[128];

    if(getc(f)==1)
    {
	n=getw(f);
	if(n>0)
	{
	    hotspot=malloc(n*sizeof(HOTSPOT));
	    if(hotspot)
	    {
		l=getw(f); // macro data size
		getw(f);   // hiword ignored
		fread(hotspot,sizeof(HOTSPOT),n,f);
		for(i=0;i<l;i++) getc(f);
		for(i=0;i<n;i++)
		{
		    for(l=0;l<sizeof(name)&&(name[l]=getc(f))!='\0';l++);
		    for(l=0;l<sizeof(buffer)&&(buffer[l]=getc(f))!='\0';l++);
		    printf("'%s' ",buffer);
		    if((hotspot[i].c1&0xF0)==0xC0)
		    {
			printf("Macro ");
		    }
		    else if(hotspot[i].c1&1)
		    {
			printf("Jump ");
		    }
		    else
		    {
			printf("Pop-up ");
		    }
		    if(hotspot[i].c2)
		    {
			printf("Invisible");
		    }
		    else
		    {
			printf("Visible ");
		    }
		    printf(" '%s' %d,%d,%d,%d\n",name,hotspot[i].x,hotspot[i].y,hotspot[i].x+hotspot[i].w,hotspot[i].y+hotspot[i].h);
		}
		free(hotspot);
	    }
	}
    }
}

int main(int argc,char *argv[])
{
    FILE *f;
    FILE *fTarget;
    int i,j,k,l,m,n,colors;
    unsigned char byType,byPacked;
    long dwDataSize,dwHotspotOffset,dwHotspotSize,w,h,offset;
    BITMAPFILEHEADER bmfh;
    BITMAPINFOHEADER bmih;
    APMFILEHEADER afh;
    char filename[PATH_MAX];
    BOOL res[7];

    assert(sizeof(HOTSPOT)==15);

    if(argc<2)
    {
	fprintf(stderr,"SPLITMRB - splits MRB into SHG/BMP/WMF and SHG into BMP/WMF files - Version 1.5\n"
		       "M.Winterhoff <mawin@gmx.net>, Geschw.-Scholl-Ring 17, 38444 Wolfsburg, Germany\n"
		       "\n"
		       "usage: SPLITMRB filename[.mrb] ...\n"
		       "       SPLITMRB filename.shg ...\n"
		       "\n"
		       "Output files are created in the current directory using resolution-dependent\n"
		       "extensions *.EGA,*.VGA,*.CGA,*.854,*.MAC or *.BMP,*.WMF or *.Snn,*.nnn (where\n"
		       "n is a digit from 0 to 9). Discarded hotspot info will be written to stdout.\n"
		       "\n");
    }
    else for(i=1;i<argc;i++)
    {
	strcpy(filename,argv[i]);
	l=strlen(filename);
	while(l>0&&filename[l-1]!='\\'&&filename[l-1]!='/'&&filename[l-1]!=':') l--;
	m=l;
	while(filename[l]!='\0'&&filename[l]!='.') l++;
	if(filename[l]=='\0') strcpy(filename+l,".mrb");
	f=fopen(filename,"rb");
	if(!f)
	{
	    fprintf(stderr,"Can not open '%s'\n",filename);
	}
	else
	{
	    if((getw(f)&0xDFFF)!=0x506C)
	    {
		fprintf(stderr,"'%s' is no MRB/SHG bitmap\n",argv[i]);
	    }
	    else
	    {
		n=getw(f);
		res[0]=res[1]=res[2]=res[3]=res[4]=res[5]=res[6]=FALSE;
		for(j=0;j<n;j++)
		{
		    fseek(f,4*(j+1),SEEK_SET);
		    fread(&offset,sizeof(offset),1,f);
		    fseek(f,offset,SEEK_SET);
		    byType=getc(f); // type of picture: 5=DDB, 6=DIB, 8=METAFILE
		    byPacked=getc(f); // packing method: 0=unpacked, 1=RunLen, 2=LZ77
		    if(byType==6||(byType==5&&byPacked<2))
		    {
			memset(&bmfh,0,sizeof(bmfh));
			memset(&bmih,0,sizeof(bmih));
			bmfh.bfType=0x4D42; // bitmap magic ("BM")
			bmih.biSize=sizeof(bmih);
			w=GetCDWord(f);
			bmih.biXPelsPerMeter=(w*79L+1)/2;
			h=GetCDWord(f);
			bmih.biYPelsPerMeter=(h*79L+1)/2;
			bmih.biPlanes=GetCWord(f);
			bmih.biBitCount=GetCWord(f);
			bmih.biWidth=GetCDWord(f);
			bmih.biHeight=GetCDWord(f);
			colors=(int)(bmih.biClrUsed=GetCDWord(f));
			if(!colors) colors=1<<bmih.biBitCount;
			bmih.biClrImportant=GetCDWord(f);
			dwDataSize=GetCDWord(f);
			dwHotspotSize=GetCDWord(f);
			GetDWord(f); // dwPictureOffset
			dwHotspotOffset=GetDWord(f);
			if(dwHotspotOffset&&n>1) // dwHotspotOffset
			{
			    // save as SHG
			    sprintf(filename+l,".S%02d",j);
			    fTarget=fopen(filename+m,"wb");
			    if(fTarget)
			    {
				fprintf(stderr,"Extracting Segmented Hotspot Graphic '%s'...\n",filename+m);
				putw(0x506C,fTarget);
				putw(1,fTarget);
				putw(8,fTarget);
				putw(0,fTarget);
				fseek(f,offset,SEEK_SET);
				copy(f,dwHotspotOffset+dwHotspotSize,fTarget);
				fclose(fTarget);
			    }
			    else
			    {
				fprintf(stderr,"Can not create '%s'\n",filename+m);
			    }
			}
			else
			{
			    sprintf(filename+l,".%03d",j);
			    if(w==96&&h==48&&!res[0])
			    {
				strcpy(filename+l,".cga");
				res[0]=TRUE;
			    }
			    else if(w==96&&h==72&&!res[1])
			    {
				strcpy(filename+l,".ega");
				res[1]=TRUE;
			    }
			    else if(w==96&&h==96&&!res[2])
			    {
				strcpy(filename+l,".vga");
				res[2]=TRUE;
			    }
			    else if(w==120&&h==120&&!res[3])
			    {
				strcpy(filename+l,".854");
				res[3]=TRUE;
			    }
			    else if(w==72&&h==72&&!res[4])
			    {
				strcpy(filename+l,".mac");
				res[4]=TRUE;
			    }
			    else if(!res[6])
			    {
				strcpy(filename+l,".bmp");
				res[6]=TRUE;
			    }
			    fTarget=fopen(filename+m,"wb");
			    if(fTarget)
			    {
				fprintf(stderr,"Extracting %ld x %ld x %u Bitmap '%s' (%ldx%ld dpi)...\n",bmih.biWidth,bmih.biHeight,bmih.biPlanes*bmih.biBitCount,filename+m,w,h);
				fwrite(&bmfh,1,sizeof(bmfh),fTarget);
				fwrite(&bmih,1,sizeof(bmih),fTarget);
				if(byType==6)
				{
				    copy(f,colors*4L,fTarget);
				}
				else
				{
				    long l;

				    l=0x000000L;
				    fwrite(&l,1,sizeof(l),fTarget);
				    l=0xFFFFFFL;
				    fwrite(&l,1,sizeof(l),fTarget);
				}
				bmfh.bfOffBits=sizeof(bmfh)+sizeof(bmih)+colors*4L;
				bmih.biSizeImage=(((bmih.biWidth*bmih.biBitCount+31)/32)*4)*bmih.biHeight;
				if(byType==5) // convert 3.0 DDB to 3.1 DIB
				{
				    long width,length,l;
				    int pad;

				    width=((bmih.biWidth*bmih.biBitCount+15)/16)*2;
				    pad=(int)(((width+3)/4)*4-width);
				    GetPackedByte(NULL);
				    for(l=0;l<bmih.biHeight;l++)
				    {
					if(byPacked==1)
					{
					    for(length=0;length<width;length++) putc(GetPackedByte(f),fTarget);
					}
					else
					{
					    copy(f,width,fTarget);
					}
					if(pad) fwrite("    ",pad,1,fTarget);
				    }
				}
				else
				{
				    decompress(byPacked,f,dwDataSize,fTarget);
				}
				// update bitmap headers
				bmfh.bfSize=ftell(fTarget);
				fseek(fTarget,0L,SEEK_SET);
				fwrite(&bmfh,1,sizeof(bmfh),fTarget);
				fwrite(&bmih,1,sizeof(bmih),fTarget);
				fclose(fTarget);
				if(dwHotspotOffset)
				{
				    fseek(f,offset+dwHotspotOffset,SEEK_SET);
				    PrintHotspotInfo(f);
				}
			    }
			    else
			    {
				fprintf(stderr,"Can not create '%s'\n",filename+m);
			    }
			}
		    }
		    else if(byType==8) // Windows MetaFile
		    {
			unsigned short *wp;

			memset(&afh,0,sizeof(afh));
			afh.dwKey=0x9AC6CDD7L;
			GetCWord(f); // mapping mode
			afh.rcBBox.right=getw(f); // width of metafile-picture
			afh.rcBBox.bottom=getw(f); // height of metafile-picture
			GetCDWord(f);
			dwDataSize=GetCDWord(f);
			dwHotspotSize=GetCDWord(f); // dwHotspotSize
			GetDWord(f); // dwPictureOffset
			dwHotspotOffset=GetDWord(f);
			if(dwHotspotOffset&&n>1) // dwHotspotOffset
			{
			    sprintf(filename+l,".S%02d",j);
			    fTarget=fopen(filename+m,"wb");
			    if(fTarget)
			    {
				fprintf(stderr,"Extracting Segmented Hotspot Graphic '%s'...\n",filename+m);
				putw(0x506C,fTarget);
				putw(1,fTarget);
				putw(8,fTarget);
				putw(0,fTarget);
				fseek(f,offset,SEEK_SET);
				copy(f,dwHotspotOffset+dwHotspotSize,fTarget);
				fclose(fTarget);
			    }
			    else
			    {
				fprintf(stderr,"Can not create '%s'\n",filename+m);
			    }
			}
			else
			{
			    afh.wInch=2540;
			    wp=(unsigned short *)&afh;
			    for(k=0;k<10;k++) afh.wChecksum^=*wp++;
			    if(!res[5])
			    {
				strcpy(filename+l,".wmf");
				res[5]=TRUE;
			    }
			    else
			    {
				sprintf(filename+l,".%03d",j);
			    }
			    fTarget=fopen(filename+m,"wb");
			    if(fTarget)
			    {
				fprintf(stderr,"Extracting Metafile '%s'...\n",filename+m);
				fwrite(&afh,1,sizeof(afh),fTarget);
				decompress(byPacked,f,dwDataSize,fTarget);
				fclose(fTarget);
			    }
			    if(dwHotspotOffset)
			    {
				fseek(f,offset+dwHotspotOffset,SEEK_SET);
				PrintHotspotInfo(f);
			    }
			}
		    }
		    else
		    {
			fprintf(stderr,"Picture %d of '%s' unknown type %02x pack=%02x skipped.\n",j,argv[i],byType,byPacked);
		    }
		}
	    }
	    fclose(f);
	}
    }
    return 0;
}
